In [12]:
import pandas as pd


In [13]:
df = pd.read_csv(
    'claybowls_starred.csv')


In [14]:
audio_features = ['danceability', 'energy', 'loudness', 'speechiness', 'acousticness',
                  'instrumentalness', 'liveness', 'valence', 'tempo']


In [15]:
for feature in audio_features:
    df['abs_diff'] = abs(df[feature] - df[feature].mean())
    most_average_song = df.loc[df['abs_diff'].idxmin()]


Calculate the mean value of the current audio feature using df[feature].mean().
Subtract the mean value from the respective feature value for each track and take the absolute difference, storing the result in a new column named 'abs_diff'.
Find the index of the minimum value in the 'abs_diff' column using df['abs_diff'].idxmin(). This index represents the most average song for the current audio feature.
Retrieve the row corresponding to the index found above using df.loc[] and store it in a variable named most_average_song.

In [16]:
print(f"Most average song based on {feature}:")
print(f"Name: {most_average_song['name']}")
print(f"Artist: {most_average_song['artist']}")
print(f"Album: {most_average_song['album']}\n")


We print the most average song information, including the song name, artist, and album, based on the current audio feature. This is done inside the loop, so it prints the result for each audio feature.

In summary, this code reads a CSV file containing track metadata, calculates the most average song for each audio feature, and prints the results.

In [17]:
# Print the average value of each audio feature
for feature in audio_features:
    average_value = df[feature].mean()
    print(f"Average {feature}: {average_value}")


### Here are the results (for Jacob's data) and the average value:
- Average danceability: 0.45695233785822015
- Average energy: 0.7061477978883861
- Average loudness: -8.068892911010558
- Average speechiness: 0.06157134238310708
- Average acousticness: 0.19087305079939668
- Average instrumentalness: 0.20570761805429869
- Average liveness: 0.22565188536953243
- Average valence: 0.4674600301659125
- Average tempo: 121.8616907993967

## Clayton's Results:
- Average danceability: 0.6540931876606684
- Average energy: 0.5704057455012853
- Average loudness: -9.419962724935733
- Average speechiness: 0.10013412596401028
- Average acousticness: 0.285686521336761
- Average instrumentalness: 0.20419770015424166
- Average liveness: 0.20164370179948587
- Average valence: 0.5648854113110541
- Average tempo: 119.99582904884319

### Now it is possible to return a list of songs that are near those values:

This code calculates the absolute difference between each audio feature and its respective average value, and then sorts the DataFrame based on that difference using the nsmallest() function. The nsmallest() function returns the ten rows with the smallest values in the 'abs_diff' column, which represent the ten songs nearest to the average value for each audio feature.

In [18]:
# Find the ten nearest songs for each audio feature
for feature in audio_features:
    average_value = df[feature].mean()
    df['abs_diff'] = abs(df[feature] - average_value)
    nearest_songs = df.nsmallest(10, 'abs_diff')

    print(f"Ten nearest songs to the average {feature} ({average_value}):")
    for idx, row in nearest_songs.iterrows():
        print(f"- {row['name']} by {row['artist']} - {feature}: {row[feature]}")
    print("- \n")

Ten nearest songs to the average danceability (0.6540931876606684):
- Float (feat. Zion I) by Minnesota - danceability: 0.655
- This Is It (Luchini) (Sedgwick Ave. Demo) by Camp Lo - danceability: 0.655
- Cocaine by Sleepwalkers - danceability: 0.655
- ReEmergence by STS9 - danceability: 0.653
- Fuck a Bomb by Hey Steve - danceability: 0.653
- Nellie Kane by Hot Rize - danceability: 0.653
- Jupiter (with Mereba & JID feat. EARTHGANG, Jurdan Bryant, Hollywood JB & Benji.) by Spillage Village - danceability: 0.653
- Time Will Tell - Live by The Magpie Salute - danceability: 0.653
- Another Day in Paradise by Quinn XCII - danceability: 0.653
- Goooo by TNGHT - danceability: 0.653
- 

Ten nearest songs to the average energy (0.5704057455012853):
- Ritmo Novo by Sango - energy: 0.57
- late nite by [bsd.u] - energy: 0.57
- Love Like This by Wild Belle - energy: 0.571
- I’m Tight by Louis Cole - energy: 0.571
- Dubbin Out by Desmond Cheese - energy: 0.569
- One For Chuck by Circles Around The

## Results

## Clayton's Results
**Ten nearest songs to the average danceability (0.6540931876606684):
- Float (feat. Zion I) by Minnesota - danceability: 0.655
- This Is It (Luchini) (Sedgwick Ave. Demo) by Camp Lo - danceability: 0.655
- Cocaine by Sleepwalkers - danceability: 0.655
- ReEmergence by STS9 - danceability: 0.653
- Fuck a Bomb by Hey Steve - danceability: 0.653
- Nellie Kane by Hot Rize - danceability: 0.653
- Jupiter (with Mereba & JID feat. EARTHGANG, Jurdan Bryant, Hollywood JB & Benji.) by Spillage Village - danceability: 0.653
- Time Will Tell - Live by The Magpie Salute - danceability: 0.653
- Another Day in Paradise by Quinn XCII - danceability: 0.653
- Goooo by TNGHT - danceability: 0.653
- 

Ten nearest songs to the average energy (0.5704057455012853):
- Ritmo Novo by Sango - energy: 0.57
- late nite by [bsd.u] - energy: 0.57
- Love Like This by Wild Belle - energy: 0.571
- I’m Tight by Louis Cole - energy: 0.571
- Dubbin Out by Desmond Cheese - energy: 0.569
- One For Chuck by Circles Around The Sun - energy: 0.569
- Sunshine by Atmosphere - energy: 0.572
- Down The Road by C2C - energy: 0.572
- I Am Dreaming by DJ Frane - energy: 0.572
- Life During Exodus by Umphrey's McGee - energy: 0.572
- 

Ten nearest songs to the average loudness (-9.419962724935733):
- It Ain't All Flowers by Sturgill Simpson - loudness: -9.42
- Watch Yo Back by Dj Smokey - loudness: -9.419
- Soul Purpose by Nightmares On Wax - loudness: -9.423
- Sorry You're Sick by Ted Hawkins - loudness: -9.411
- Dance All Night by Tnertle - loudness: -9.405
- Under My Sensi by Barrington Levy - loudness: -9.403
- Astrovan by Mt. Joy - loudness: -9.4
- Slow Funk by Buddy Rich - loudness: -9.44
- African Daughter Dub by Scientist - loudness: -9.396
- Like a Feather by BoomBox - loudness: -9.39
- 

Ten nearest songs to the average speechiness (0.10013412596401028):
- Fantasy by Kari Faux - speechiness: 0.0997
- Warp by EOTO - speechiness: 0.0997
- Roast Fish And Cornbread - Extended Mix (7" Single) by Lee "Scratch" Perry - speechiness: 0.0995
- War Of The Currents by Gramatik - speechiness: 0.0993
- When a Woman's Fed Up by R. Kelly - speechiness: 0.101
- Day-O (The Banana Boat Song) - Live by Harry Belafonte - speechiness: 0.101
- Something About Us by Daft Punk - speechiness: 0.0986
- I'd Rather Go Blind by The Frightnrs - speechiness: 0.0986
- Angels and Albatrosses by Quantic - speechiness: 0.0985
- Multi-Family Garage Sale (Bargain-Bin Mix) by Land of the Loops - speechiness: 0.0979
- 

Ten nearest songs to the average acousticness (0.285686521336761):
- Handcuffs by Parliament - acousticness: 0.286
- Smokestack Lightning by Marty O'Reilly & the Old Soul Orchestra - acousticness: 0.286
- Free by Phish - acousticness: 0.285
- Tearin' It Up by Jammin Joe Buck - acousticness: 0.287
- Too Much Information by Railroad Earth - acousticness: 0.283
- Never Been To Spain by Elvis Presley - acousticness: 0.283
- Hurt Nobody (feat. MPA Wicced) by Gucci Mane - acousticness: 0.282
- How It Felt by Ryan Hemsworth - acousticness: 0.282
- Limestone by Magic City Hippies - acousticness: 0.282
- Don't Look Back by Blueprint - acousticness: 0.282
- 

Ten nearest songs to the average instrumentalness (0.20419770015424166):
- Here on Out by The Magic Beans - instrumentalness: 0.204
- Gentleman by Fela Kuti - instrumentalness: 0.205
- For The Lost Souls by Samsara Blues Experiment - instrumentalness: 0.203
- The Virus by The Halluci Nation - instrumentalness: 0.202
- Back to Nature by Nightmares On Wax - instrumentalness: 0.202
- The Virus by The Halluci Nation - instrumentalness: 0.202
- Meet Me In The Middle - From The "Fifty Shades Of Grey" Soundtrack by Jessie Ware - instrumentalness: 0.209
- Meet Me In The City by The Black Keys - instrumentalness: 0.199
- Camelblues by Mndsgn - instrumentalness: 0.198
- Sanctify by St. Paul & The Broken Bones - instrumentalness: 0.198
- 

Ten nearest songs to the average liveness (0.20164370179948587):
- Pass The Pussy by The Dogs - liveness: 0.202
- Renaissance Forever by Natia - liveness: 0.202
- Stuck In A Moment You Can't Get Out Of by U2 - liveness: 0.201
- Them Tulsa Boys by Paul Benjaman Band - liveness: 0.201
- Bicycling in Bombay by Garaj Mahal - liveness: 0.201
- Take Me Back To Tulsa by Bob Wills & His Texas Playboys - liveness: 0.201
- Nellie Kane by Hot Rize - liveness: 0.201
- Time's A Wastin by Erykah Badu - liveness: 0.201
- Droogs by NxWorries - liveness: 0.203
- Agorinha by Sango - liveness: 0.2
- 

Ten nearest songs to the average valence (0.5648854113110541):
- Like Water by Abstract Truth - valence: 0.565
- In The Rain by The Dramatics - valence: 0.564
- Serious Drug by Wildcookie - valence: 0.564
- Leaving The Nest (It's A Long Way Down) by Benjy Ferree - valence: 0.566
- UMI Says by Mos Def - valence: 0.566
- Floating to Nowhere by Lando Chill - valence: 0.566
- Time Will Tell - Live by The Magpie Salute - valence: 0.566
- Minds Awake by Rumspringa - valence: 0.563
- Black Cow by Steely Dan - valence: 0.563
- Borderline with My Atoms by Hiatus Kaiyote - valence: 0.563
- 

Ten nearest songs to the average tempo (119.99582904884319):
- Stormur by Sigur Rós - tempo: 119.997
- Talkin' To The Moon by Pauline London - tempo: 120.001
- Anybody out There by Organized Noize - tempo: 119.989
- Highway One by The Flashbulb - tempo: 119.988
- Joyful Girl by Soulive - tempo: 120.005
- Black Synagogue by Angel Haze - tempo: 120.006
- Fruit by ABRA - tempo: 120.008
- Death by Supertask - tempo: 120.011
- Kirra by CharlestheFirst - tempo: 120.014
- Extract of a Fragment Part I by Prisma - tempo: 120.015
- **

### (jacob's data)

Ten nearest songs to the average danceability (0.45695233785822015):
- Shake Me Down by Cage The Elephant - danceability: 0.457
- Today - 2011 Remaster by The Smashing Pumpkins - danceability: 0.456
- All We Have Is Now by Royal Blood - danceability: 0.459
- Little Monster by Royal Blood - danceability: 0.454
- Suite: Judy Blue Eyes by Crosby, Stills & Nash - danceability: 0.454
- Romany Dagger by All Them Witches - danceability: 0.454
- Ezy Ryder by Jimi Hendrix - danceability: 0.46
- Castles Made of Sand by Jimi Hendrix - danceability: 0.46
- Elite by Deftones - danceability: 0.46
- Afterimage by Rush - danceability: 0.453
- 

Ten nearest songs to the average energy (0.7061477978883861):
- Little Black Submarines - Radio Edit [radio edit] by The Black Keys - energy: 0.706
- Ride Like the Wind by Christopher Cross - energy: 0.707
- Help Me, Rhonda by The Beach Boys - energy: 0.707
- Ugo by The Dead Pirates - energy: 0.707
- Garden by Pearl Jam - energy: 0.707
- Angel by Jimi Hendrix - energy: 0.705
- Tighten Up by The Black Keys - energy: 0.705
- Suicidal Dream by Silverchair - energy: 0.708
- Charles William by All Them Witches - energy: 0.704
- Unity by The Comet Is Coming - energy: 0.709
- 

Ten nearest songs to the average loudness (-8.068892911010558):
- Circumstances by Rush - loudness: -8.055
- Sludge Factory - Live at the Majestic Theatre, Brooklyn, NY - April 1996 by Alice In Chains - loudness: -8.046
- The River of Dreams by Billy Joel - loudness: -8.101
- The Analog Kid by Rush - loudness: -8.034
- Over The Falls by Primus - loudness: -8.022
- Third Stone From The Sun by Jimi Hendrix - loudness: -8.119
- Incinerate by Sonic Youth - loudness: -8.016
- Love or Confusion - BBC Sessions by Jimi Hendrix - loudness: -8.122
- When God Comes Back by All Them Witches - loudness: -8.014
- Anthem by Rush - loudness: -8.124
- 

Ten nearest songs to the average speechiness (0.06157134238310708):
- Atomic Mind by Nick Johnston - speechiness: 0.0616
- Deep by Pearl Jam - speechiness: 0.0616
- How Soon Is Now? - 2011 Remaster by The Smiths - speechiness: 0.0615
- You Want Me by Royal Blood - speechiness: 0.0613
- Angel by Jimi Hendrix - speechiness: 0.0619
- Madman - Live at the Cambridge by Silverchair - speechiness: 0.0619
- Oblivion by Royal Blood - speechiness: 0.0612
- Faultline - Live at the Cambridge by Silverchair - speechiness: 0.0612
- Neodämmerung by Don Davis - speechiness: 0.062
- Israel's Son by Silverchair - speechiness: 0.0621
- 

Ten nearest songs to the average acousticness (0.19087305079939668):
- 50 Ways to Leave Your Lover by Paul Simon - acousticness: 0.19
- Schizophrenia by Sonic Youth - acousticness: 0.19
- Interlude with Ludes by Them Crooked Vultures - acousticness: 0.193
- Top Yourself by The Raconteurs - acousticness: 0.188
- Don't Eat The Yellow Snow by Frank Zappa - acousticness: 0.187
- Loose Change by Royal Blood - acousticness: 0.195
- Angel by Jimi Hendrix - acousticness: 0.186
- Do I Wanna Know? by Arctic Monkeys - acousticness: 0.186
- New World Man by Rush - acousticness: 0.185
- Wendy by The Beach Boys - acousticness: 0.183
- 

Ten nearest songs to the average instrumentalness (0.20570761805429869):
- Digital Bath by Deftones - instrumentalness: 0.211
- Findaway by Silverchair - instrumentalness: 0.212
- If Only by Queens of the Stone Age - instrumentalness: 0.195
- The Killer Is Me - Live at the Majestic Theatre, Brooklyn, NY - April 1996 by Alice In Chains - instrumentalness: 0.195
- Nutshell - Live at the Majestic Theatre, Brooklyn, NY - April 1996 by Alice In Chains - instrumentalness: 0.217
- Not a Through Street by Anya Marina - instrumentalness: 0.194
- Mary The Ice Cube by Primus - instrumentalness: 0.218
- Pretty by Korn - instrumentalness: 0.192
- Cubensis Lenses by Psychedelic Porn Crumpets - instrumentalness: 0.187
- New World Man by Rush - instrumentalness: 0.182
- 

Ten nearest songs to the average liveness (0.22565188536953243):
- Diamonds on the Soles of Her Shoes by Paul Simon - liveness: 0.224
- Antimatter Man by Man Or Astro-Man? - liveness: 0.224
- Time In A Bottle by Jim Croce - liveness: 0.228
- Reckless Tongue by Airways - liveness: 0.228
- Part III by Crumb - liveness: 0.228
- Mind Eraser, No Chaser by Them Crooked Vultures - liveness: 0.222
- Super Zodiac by The Comet Is Coming - liveness: 0.222
- Stone Free by Jimi Hendrix - liveness: 0.221
- Cornflake by Psychedelic Porn Crumpets - liveness: 0.219
- Disco Fever by Goat - liveness: 0.219
- 

Ten nearest songs to the average valence (0.4674600301659125):
- Would? - Live at the Majestic Theatre, Brooklyn, NY - April 1996 by Alice In Chains - valence: 0.467
- Mister Sandman by Chet Atkins - valence: 0.468
- Wendy by The Beach Boys - valence: 0.469
- The Analog Kid by Rush - valence: 0.465
- The Making Of by The Bohicas - valence: 0.464
- The Switch and the Spur by The Raconteurs - valence: 0.464
- Marathon by Rush - valence: 0.471
- Foxey Lady - Live In Maui, 1970 by Jimi Hendrix - valence: 0.471
- Wynona's Big Brown Beaver by Primus - valence: 0.471
- That's Amore - 1996 Digital Remaster by Dean Martin - valence: 0.463
- 

Ten nearest songs to the average tempo (121.8616907993967):
- Bullet With Butterfly Wings - Remastered 2012 by The Smashing Pumpkins - tempo: 121.864
- Nutshell - Live at the Majestic Theatre, Brooklyn, NY - April 1996 by Alice In Chains - tempo: 121.801
- One Minute To Midnight by Justice - tempo: 121.974
- The Killer Is Me - Live at the Majestic Theatre, Brooklyn, NY - April 1996 by Alice In Chains - tempo: 121.983
- Acid Rain by Silverchair - tempo: 121.74
- New York State of Mind by Billy Joel - tempo: 122.015
- Starlight by Muse - tempo: 121.61
- Sea Of Sorrow by Alice In Chains - tempo: 122.147
- The Bold Arrow Of Time by Tame Impala - tempo: 122.155
- Cicada by Silverchair - tempo: 122.156
- 

In [ ]:
# Importing necessary libraries for K-NN
from sklearn.neighbors import KNeighborsClassifier
import numpy as np
import pandas as pd

# Sample data: Replace this with your actual audio features data
X_train = np.random.rand(100, 5)  # 100 songs, 5 audio features each
y_train = np.random.randint(0, 2, 100)  # Labels (not necessary for unsupervised K-NN)
X_test = np.random.rand(10, 5)  # 10 songs to find neighbors for

# Initialize K-NN classifier
knn = KNeighborsClassifier(n_neighbors=10, metric='euclidean')

# Fit the model
knn.fit(X_train, y_train)

# Find the nearest neighbors
distances, indices = knn.kneighbors(X_test)

# Display the indices of 10 nearest songs for each song in X_test
print('Indices of 10 nearest songs:', indices)

In [ ]:
# Importing the CSV file containing Clayton's starred songs and their audio features
df = pd.read_csv('claybowls_starred.csv')

# Selecting only the audio features from the DataFrame
audio_features = ['danceability', 'energy', 'loudness', 'speechiness', 'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo']
X_train_clayton = df[audio_features].values

# Initialize K-NN classifier with Clayton's data
knn_clayton = KNeighborsClassifier(n_neighbors=10, metric='euclidean')

# Fit the model with Clayton's data
knn_clayton.fit(X_train_clayton, np.zeros(X_train_clayton.shape[0]))  # Labels set to zero as they are not necessary for unsupervised K-NN

# Clayton's average audio feature values
clayton_avg_features = np.array([0.6541, 0.5704, -9.4199, 0.1001, 0.2857, 0.2042, 0.2016, 0.5649, 119.9958])

# Reshape the array to make it compatible with scikit-learn methods
clayton_avg_features = clayton_avg_features.reshape(1, -1)

# Use K-NN to find the 10 nearest songs to Clayton's average audio features
distances, indices = knn_clayton.kneighbors(clayton_avg_features)

# Display the indices of the 10 nearest songs to Clayton's average
print('Indices of 10 nearest songs to Clayton\'s average:', indices)

In [ ]:
# Retrieve the song names corresponding to the indices of the 10 nearest songs to Clayton's average
nearest_songs = df.iloc[indices[0]]

# Display the names, artists, and albums of the 10 nearest songs
nearest_songs[['name', 'artist', 'album']]

In [ ]:
# Importing necessary libraries for data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Set the style for the plots
sns.set(style='whitegrid')

# Create a DataFrame containing only the 10 nearest songs and their audio features
nearest_songs_features = nearest_songs[audio_features]

# Add Clayton's average audio features as a new row to the DataFrame
nearest_songs_features.loc['Clayton_Avg'] = clayton_avg_features[0]

# Create a heatmap to visualize the audio features of the 10 nearest songs compared to Clayton's average
plt.figure(figsize=(12, 8))
sns.heatmap(nearest_songs_features, annot=True, cmap='coolwarm', cbar=True, xticklabels=audio_features, yticklabels=nearest_songs_features.index)
plt.title('Heatmap of Audio Features for 10 Nearest Songs Compared to Clayton\'s Average')
plt.xlabel('Audio Features')
plt.ylabel('Song Indices / Clayton\'s Average')
plt.show()